### PR TITLE
380 unentitying - REVIEW, DO NOT MERGE

### PR DESCRIPTION
**Pull request by DaanVanYperen**

### CODE REVIEW

File: artemis/src/main/java/com/artemis/EntityManager.java [hunk 1]
```diff
@@ -91,35 +103,25 @@ void updateCompositionIdentity(EntityEdit edit) {
 	int compositionIdentity(BitSet componentBits) {
 		int identity = identityResolver.getIdentity(componentBits);
 		if (identity > highestSeenIdentity) {
-			subscriptionManager.processComponentIdentity(identity, componentBits);
+			world.getSystem(AspectSubscriptionManager.class)
```
Comment (DaanVanYperen) [inline artemis/src/main/java/com/artemis/EntityManager.java]: We do this everywhere, deserves its own getter on world. We pre-resolve it into a field anyway.

File: artemis/src/main/java/com/artemis/EntityManager.java [hunk 2]
```diff
@@ -181,7 +179,7 @@ void synchronize(EntitySubscription es) {
 		for (int i = 0; i < entities.size(); i++) {
 			Entity e = entities.get(i);
 			if (e != null)
-				es.check(e.id);
+				es.check(e.getId());
```
Comment (DaanVanYperen) [inline artemis/src/main/java/com/artemis/EntityManager.java]: wish we made up our minds :P

File: artemis/src/main/java/com/artemis/InvocationStrategy.java [hunk 3]
```diff
@@ -7,11 +7,10 @@
 	protected void process(Bag<BaseSystem> systems) {
 		Object[] systemsData = systems.getData();
 		for (int i = 0, s = systems.size(); s > i; i++) {
-			updateEntityStates();
-
 			BaseSystem system = (BaseSystem) systemsData[i];
 			if (!system.isPassive()) {
 				system.process();
+				updateEntityStates();
```
Comment (DaanVanYperen) [inline artemis/src/main/java/com/artemis/InvocationStrategy.java]: workflow change. (add to changelog)

File: artemis/src/main/java/com/artemis/Manager.java [hunk 4]
```diff
@@ -9,38 +11,41 @@
...
+				});
 	}
 
-	@Override
-	public final void changed(IntBag entities) {
+	private void added(IntBag entities) {
 		int[] ids = entities.getData();
 		for (int i = 0, s = entities.size(); s > i; i++) {
-			changed(ids[i]);
+			added(world.getEntity(ids[i]));
```
Comment (DaanVanYperen) [inline artemis/src/main/java/com/artemis/Manager.java]: Should definitely discourage using Manager in javadoc.

File: artemis/src/main/java/com/artemis/Manager.java [hunk 5]
```diff
@@ -9,38 +11,41 @@
...
+
+					@Override
+					public void removed(IntBag entities) {
+						deleted(entities);
+					}
+				});
 	}
 
-	@Override
-	public final void changed(IntBag entities) {
```
Comment (DaanVanYperen) [inline artemis/src/main/java/com/artemis/Manager.java]: Are we handicapping any users by removing changed?

File: artemis/src/main/java/com/artemis/WorldConfiguration.java [hunk 6]
```diff
@@ -174,6 +174,9 @@ void initialize(World world, Injector injector, AspectSubscriptionManager asm) {
 		for (BaseSystem system : systems) {
 			world.systems.put(system.getClass(), system);
 			system.setWorld(world);
+			if (system instanceof Manager) {
```
Comment (DaanVanYperen) [inline artemis/src/main/java/com/artemis/WorldConfiguration.java]: *cough*

File: artemis/src/main/java/com/artemis/managers/TagManager.java [hunk 7]
```diff
@@ -17,21 +20,53 @@
  * 
  * @author Arni Arent
  */
-public class TagManager extends Manager {
+public class TagManager extends BaseSystem {
```
Comment (DaanVanYperen) [inline artemis/src/main/java/com/artemis/managers/TagManager.java]: todo dy: move state to component.

File: artemis/src/main/java/com/artemis/managers/TagManager.java [hunk 8]
```diff
@@ -47,6 +82,11 @@ public TagManager() {
 	public void register(String tag, Entity e) {
 		entitiesByTag.put(tag, e);
 		tagsByEntity.put(e, tag);
+		registered.set(e.getId());
+	}
+
+	public void register(String tag, int entityId) {
+		register(tag, world.getEntity(entityId));
```
Comment (DaanVanYperen) [inline artemis/src/main/java/com/artemis/managers/TagManager.java]: I see what you mean by cheating. ;) Should be fine for now.

**Comment by DaanVanYperen** [comment]
Since the goal is old API, what about enable/disable and UUID? Think we can write off UUID as bad practice. Thinking we should leave as is and cover it with a migration guide, with practical example implementations for both.

File: artemis/src/main/java/com/artemis/EntityManager.java (hunk 2 above)
Comment (junkdog) [inline artemis/src/main/java/com/artemis/EntityManager.java]: `Entity#id` is package local though - it's mostly about disallowing users from updating the id.

File: artemis/src/main/java/com/artemis/InvocationStrategy.java (hunk 3 above)
Comment (junkdog) [inline artemis/src/main/java/com/artemis/InvocationStrategy.java]: No change - `updateEntityStates()` is populated during system.process - if systems don't run, there can't be any entity changes to update.

File: artemis/src/main/java/com/artemis/Manager.java (hunk 4 above)
Comment (junkdog) [inline artemis/src/main/java/com/artemis/Manager.java]: Perf impact isn't so bad, actually - it''s cheaper than processing systems; I left the optimization opportunity in there though.

File: artemis/src/main/java/com/artemis/Manager.java (hunk 5 above)
Comment (junkdog) [inline artemis/src/main/java/com/artemis/Manager.java]: > Are we handicapping any users by removing changed?

I thought about it, but SubscriptionListeners with proper aspects should cover all use-cases.  "Should".

>  Also consider people using EntityObserver outside of systems.

EntityObservers have been strictly used by Managers though - it's only 1.0.0 which was a bit more lenient about it.

**Comment by junkdog** [comment]
> What about enable/disable and UUID convenience methods?

Enable/disable - can be emulated with future default Aspects, the underlying code is gone. I don't think it's a very commonly used feature? It's gemserk's pooling hack which got pulled in by vanilla artemis.

> UUID convenience 

I get the feeling people are more inclined thinking about using it than actually using it. It's hard seeing where UUID is the right choice.

**Comment by DaanVanYperen** [comment]
https://github.com/junkdog/artemis-odb/wiki/Upgrading-Checklist-1.0.0

File: artemis/src/main/java/com/artemis/InvocationStrategy.java (hunk 3 above)
Comment (DaanVanYperen) [inline artemis/src/main/java/com/artemis/InvocationStrategy.java]: We've had at least one user calling systems outside of this scope, some people are free spirits! You're right though, implementation.

File: artemis/src/main/java/com/artemis/Manager.java (hunk 5 above)
Comment (DaanVanYperen) [inline artemis/src/main/java/com/artemis/Manager.java]: ack, lets wait for edge cases to report themselves for 1.1

File: artemis/src/main/java/com/artemis/InvocationStrategy.java (hunk 3 above)
Comment (junkdog) [inline artemis/src/main/java/com/artemis/InvocationStrategy.java]: hmm... but it still requires calling updateEntityStates separately, if the system create/edits/deletes entities. Tempting to deprecate passive() - it should be handled by the invocation strategy instead. Would mess with a lot of code bases though.

File: artemis/src/main/java/com/artemis/managers/TagManager.java (hunk 7 above)
Comment (junkdog) [inline artemis/src/main/java/com/artemis/managers/TagManager.java]: I was against it in the past, but the more I think about it, the more sense it makes

Comment (DaanVanYperen) [inline artemis/src/main/java/com/artemis/managers/TagManager.java]: I pay Linda to whisper instructions into your ear every night. :dancers:

**Pull request merged**